### PR TITLE
Feature/composition playlist canonical model

### PR DESCRIPTION
**Pull request by schakrovorthy**

### CODE REVIEW

File: src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java [hunk 1]
```diff
@@ -67,14 +66,35 @@
...
+            return this.pklSchemaPath;
+        }
+
+        private String getPKLContext(){
+            return this.pklContext;
+        }
+    }
+    public static final List<PKLSchema> supportedPKLSchemas = Collections.unmodifiableList
+            (new ArrayList<PKLSchema>() {{ add( new PKLSchema("org/smpte_ra/schemas/st0429_8_2007/PKL/packingList_schema.xsd", "org.smpte_ra.schemas.st0429_8_2007.PKL"));
+                add( new PKLSchema("org/smpte_ra/schemas/st0429_8_2016/PKL/packingList_schema.xsd", "org.smpte_ra.schemas.st0429_8_2016.PKL"));}});
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java]: Given that the namespace for 2016 PKL is "http://www.smpte-ra.org/schemas/2067-2/2016/PKL", should we consider renaming arguments "org/smpte_ra/schemas/st0429_8_2016/PKL/packingList_schema.xsd" and "org.smpte_ra.schemas.st0429_8_2016.PKL"?

Comment (schakrovorthy) [inline src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java]: Ofcourse, will modify the variable names and submit another commit.

File: src/main/java/com/netflix/imflibrary/imp_validation/IMFMasterPackage.java [hunk 2]
```diff
@@ -61,7 +52,7 @@ public IMFMasterPackage(List<ResourceByteRangeProvider> resourceByteRangeProvide
                 assetMaps.add(resourceByteRangeProvider);
             } else if (PackingList.isFileOfSupportedSchema(resourceByteRangeProvider)) {
                 packingListsResourceByteRangeProviders.add(resourceByteRangeProvider);
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/imp_validation/IMFMasterPackage.java]: Should we consider renaming 'packingListsResourceByteRangeProviders' to 'packingLists'?

File: src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java [hunk 3]
```diff
@@ -138,44 +120,121 @@ public PackingList(ResourceByteRangeProvider resourceByteRangeProvider, @Nonnull
...
+                //this.packingListType = PackingList.checkConformance(packingListTypeJAXBElement.getValue());
+                org.smpte_ra.schemas.st0429_8_2007.PKL.PackingListType packingListType = (org.smpte_ra.schemas.st0429_8_2007.PKL.PackingListType) this.packingListTypeJAXBElement.getValue();
+                this.uuid = UUIDHelper.fromUUIDAsURNStringToUUID(packingListType.getId());
+
+                for (org.smpte_ra.schemas.st0429_8_2007.PKL.AssetType assetType : packingListType.getAssetList().getAsset()) {
+                    Asset asset = new Asset(assetType.getId(), Arrays.copyOf(assetType.getHash(), assetType.getHash().length), assetType.getSize().longValue(), assetType.getType(), assetType.getOriginalFileName().getValue());
+                    this.assetList.add(asset);
+                }
+                break;
+            case "org.smpte_ra.schemas.st0429_8_2016.PKL":
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java]: The string in question in the case statement appears to be "org.smpte_ra.schemas.st2067_2_2016.PKL"

File: src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java [hunk 4]
```diff
@@ -138,44 +120,121 @@ public PackingList(ResourceByteRangeProvider resourceByteRangeProvider, @Nonnull
...
+                throw new IMFException(String.format("Please check the PKL document and namespace URI, currently we only support the 2007 PKL schema URI"));
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java]: i might be having a memory lapse here - is there a reason why do not have an implementation for the 2016 case? Is it because we have not generated the model yet?

File: src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java [hunk 5]
```diff
@@ -68,16 +68,35 @@
...
     private final List<Asset> packingListAssets = new ArrayList<>();
     private final Map<UUID, URI> uuidToPath = new HashMap<>();
     private static final Logger logger = LoggerFactory.getLogger(AssetMap.class);
-    private final AssetMapType assetMapType;
+    private final JAXBElement assetMapTypeJAXBElement;
     public static final List<String> supportedAssetMapSchemaURIs = Collections.unmodifiableList(new ArrayList<String>(){{ add("http://www.smpte-ra.org/schemas/429-9/2007/AM");}});
 
+    public static final List<AssetMapSchema> supportedAssetMapSchemas = Collections.unmodifiableList
+            (new ArrayList<AssetMapSchema>() {{ add( new AssetMapSchema("/org/smpte_ra/schemas/st0429_9_2007/AM/assetMap_schema.xsd", "org.smpte_ra.schemas.st0429_9_2007.AM"));
+                add( new AssetMapSchema("/org/smpte_ra/schemas/st0429_9_2016/AM/assetMap_schema.xsd", "org.smpte_ra.schemas.st0429_9_2016.AM"));}});
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java]: AFAIK, there exists only one schema for AssetMap? Please let me know if that is not correct

File: src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java [hunk 6]
```diff
@@ -106,57 +125,72 @@ public AssetMap(ResourceByteRangeProvider resourceByteRangeProvider, @Nonnull IM
...
+
+        for(int i=0; i<supportedAssetMapSchemas.size(); i++) {
+            try (InputStream inputStream = resourceByteRangeProvider.getByteRangeAsStream(0, resourceByteRangeProvider.getResourceSize() - 1);
+                 InputStream assetMap_schema_is = AssetMap.class.getResourceAsStream(supportedAssetMapSchemas.get(i).getAssetMapSchemaPath());) {
+                SchemaFactory schemaFactory = SchemaFactory.newInstance(XMLConstants.W3C_XML_SCHEMA_NS_URI);
+                StreamSource schemaSource = new StreamSource(assetMap_schema_is);
+                Schema schema = schemaFactory.newSchema(schemaSource);
+
+                ValidationEventHandlerImpl validationEventHandlerImpl = new ValidationEventHandlerImpl(true);
+                JAXBContext jaxbContext = JAXBContext.newInstance("org.smpte_ra.schemas.st0429_9_2007.AM");
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java]: Is there a hard-coding typo in the parameter passed to JAXBContext?

File: src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java [hunk 7]
```diff
@@ -106,57 +125,72 @@ public AssetMap(ResourceByteRangeProvider resourceByteRangeProvider, @Nonnull IM
...
-            JAXBElement<AssetMapType> assetMapTypeJAXBElement = (JAXBElement)unmarshaller.unmarshal(inputStream);
-            if(validationEventHandlerImpl.hasErrors())
-            {
-                List<ValidationEventHandlerImpl.ValidationErrorObject> errors = validationEventHandlerImpl.getErrors();
-                for(ValidationEventHandlerImpl.ValidationErrorObject error : errors){
-                    imfErrorLogger.addError(IMFErrorLogger.IMFErrors.ErrorCodes.IMF_AM_ERROR, error.getValidationEventSeverity(), error.getErrorMessage());
+        JAXBElement assetMapTypeJAXBElement = null;
+        AssetMapSchema assetMapSchema = null;
+
+        for(int i=0; i<supportedAssetMapSchemas.size(); i++) {
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st0429_9/AssetMap.java]: Can any inference from "validateAssetMapSchema()" be used to avoid looping over supported schemas again?

File: src/main/java/com/netflix/imflibrary/st2067_2/Composition.java [hunk 8]
```diff
@@ -0,0 +1,1154 @@
...
+        return result;
+    }
+
+    private final String getCPLNamespaceVersion(String namespaceURI){
+        String[] uriComponents = namespaceURI.split("/");
+        String namespaceVersion = uriComponents[uriComponents.length - 1];
+        return namespaceVersion;
+    }
+
+    private final String serializeCollectionToString(Collection<String> set){
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Composition.java]: Does this method belong to another class perhaps in 'com.netflix.imflibrary.utils' package? Can it be genericized?

File: src/main/java/com/netflix/imflibrary/st2067_2/Composition.java [hunk 9]
```diff
@@ -0,0 +1,1154 @@
...
+    public UUID getUUID()
+    {
+        return this.uuid;
+    }
+
+    /**
+     * Getter for the CompositionPlaylistType object model of the Composition defined by the st2067-3 schema.
+     * @return the composition playlist type object model.
+     */
+    public JAXBElement getCompositionPlaylistTypeJAXBElement(){
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Composition.java]: Wouldnt it be better if this was an internal method of the class? e.g., that would help hide the JAXB aspects of our implementation

File: src/main/java/com/netflix/imflibrary/st2067_2/CompositionModels/IMFCoreConstraintsChecker_st2067_2_2013.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * A class that performs 2013 CoreConstraints st2067-2:2013 related checks on the elements of a Composition Playlist such as VirtualTracks, Segments, Sequences and Resources.
+ */
+public final class IMFCoreConstraintsChecker_st2067_2_2013 {
```
Comment (rpuri-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/CompositionModels/IMFCoreConstraintsChecker_st2067_2_2013.java]: It might be a good idea to add a private constructor to this class whose body is empty - to prevent instantiation
